proof/refine/AARCH64/TcbAcc_R.thy

```diff
@@ -4390,12 +5311,416 @@ lemma sts_utr[wp]:
+  "tcbSchedEnqueue tcbPtr \<lbrace>valid_sched_pointers\<rbrace>"
+  apply (clarsimp simp: tcbSchedEnqueue_def getQueue_def unless_def)
+  \<comment> \<open>we step forwards until we can step over the addToBitmap in order to avoid state blow-up\<close>
+  apply (intro hoare_seq_ext[OF _ stateAssert_sp] hoare_seq_ext[OF _ isRunnable_inv]
+               hoare_seq_ext[OF _ assert_sp] hoare_seq_ext[OF _ threadGet_sp]
+               hoare_seq_ext[OF _ gets_sp]
+         | rule hoare_when_cases, fastforce)+
+  apply (forward_inv_step wp: hoare_vcg_ex_lift)
+  supply if_split[split del]
+  apply (wpsimp wp: getTCB_wp
+              simp: threadSet_def setObject_def updateObject_default_def tcbQueuePrepend_def
+                    setQueue_def)
+  apply (clarsimp simp: valid_sched_pointers_def)
+  apply (intro conjI impI)
+   apply (fastforce simp: opt_pred_def opt_map_def split: if_splits)
+  apply normalise_obj_at'
+  apply (clarsimp simp: ksReadyQueues_asrt_def)
+  apply (drule_tac x="tcbDomain tcb" in spec)
+  apply (drule_tac x="tcbPriority tcb" in spec)
+  apply (clarsimp simp: valid_sched_pointers_def list_queue_relation_def)
+  apply (case_tac "ts = []", fastforce simp: tcbQueueEmpty_def)
+  by (intro conjI impI;
+      force dest!: hd_in_set heap_path_head
+             simp: inQ_def opt_pred_def opt_map_def obj_at'_def split: if_splits)
+
+lemma tcbSchedAppend_valid_sched_pointers[wp]:
+  "tcbSchedAppend tcbPtr \<lbrace>valid_sched_pointers\<rbrace>"
+  apply (clarsimp simp: tcbSchedAppend_def getQueue_def unless_def)
+  \<comment> \<open>we step forwards until we can step over the addToBitmap in order to avoid state blow-up\<close>
+  apply (intro hoare_seq_ext[OF _ stateAssert_sp] hoare_seq_ext[OF _ isRunnable_inv]
+               hoare_seq_ext[OF _ assert_sp] hoare_seq_ext[OF _ threadGet_sp]
+               hoare_seq_ext[OF _ gets_sp]
+         | rule hoare_when_cases, fastforce)+
+  apply (forward_inv_step wp: hoare_vcg_ex_lift)
+  supply if_split[split del]
+  apply (wpsimp wp: getTCB_wp
+              simp: threadSet_def setObject_def updateObject_default_def tcbQueueAppend_def
+                    setQueue_def)
+  apply (clarsimp simp: valid_sched_pointers_def)
+  apply (intro conjI impI)
+   apply (fastforce simp: opt_pred_def opt_map_def split: if_splits)
+  apply normalise_obj_at'
+  apply (clarsimp simp: ksReadyQueues_asrt_def)
+  apply (drule_tac x="tcbDomain tcb" in spec)
+  apply (drule_tac x="tcbPriority tcb" in spec)
+  by (intro conjI impI;
+      clarsimp dest: last_in_set
+               simp: valid_sched_pointers_def opt_map_def list_queue_relation_def tcbQueueEmpty_def
+                     queue_end_valid_def inQ_def opt_pred_def obj_at'_def
+              split: if_splits option.splits;
+      fastforce)
+
+(* FIXME: find a suitable place for the following lemmas regarding setObject *)
+lemma koTypeOf_injectKO:
+  fixes v :: "'a :: pspace_storable" shows
+  "koTypeOf (injectKO v) = koType TYPE('a)"
```

Comment:
These should usually go into KHeap_R or maybe Bits_R

Reply:
Moved to the very top of `KHeap_R`